samza-core/src/main/java/org/apache/samza/clustermanager/StandbyContainerManager.java

```diff
@@ -0,0 +1,466 @@
+      AbstractContainerAllocator containerAllocator) {
+    Optional<Entry<String, SamzaResource>> standbyContainer = this.selectStandby(containerID, resourceID);
+
+    // If we find a standbyContainer, we initiate a failover
+    if (standbyContainer.isPresent()) {
+
+      // ResourceID of the active container at the time of its last failure
+      String standbyContainerId = standbyContainer.get().getKey();
+      SamzaResource standbyResource = standbyContainer.get().getValue();
+      String standbyResourceID = standbyResource.getResourceID();
+      String standbyHost = standbyResource.getHost();
+
+      // update the failover state
+      this.registerFailover(containerID, resourceID, standbyResourceID, standbyHost);
+      log.info("Initiating failover and stopping standby container, found standbyContainer {} = resource {}, "
+          + "for active container {}", standbyContainerId, standbyResourceID, containerID);
+      samzaApplicationState.failoversToStandby.incrementAndGet();
+
+      this.clusterResourceManager.stopStreamProcessor(standbyResource);
+      return;
+    } else {
+
+      // If we dont find a standbyContainer, we proceed with the ANYHOST request
+      log.info("No standby container found for active container {}, making a request for {}", containerID,
+          ResourceRequestState.ANY_HOST);
+      samzaApplicationState.failoversToAnyHost.incrementAndGet();
+      containerAllocator.requestResource(containerID, ResourceRequestState.ANY_HOST);
+      return;
+    }
+  }
+
+  /**
+   * Method to select a standby container for a given active container that has stopped.
+   * TODO: enrich this method to select standby's intelligently based on lag, timestamp, load-balencing, etc.
+   * @param activeContainerID Samza containerID of the active container
+   * @param activeContainerResourceID ResourceID of the active container at the time of its last failure
+   * @return
+   */
+  private Optional<Entry<String, SamzaResource>> selectStandby(String activeContainerID,
+      String activeContainerResourceID) {
+
+    log.info("Standby containers {} for active container {}", this.standbyContainerConstraints.get(activeContainerID), activeContainerID);
+
+    // obtain any existing failover metadata
+    Optional<StandbyContainerManager.FailoverMetadata> failoverMetadata =
+        activeContainerResourceID == null ? Optional.empty() : this.getFailoverMetadata(activeContainerResourceID);
+
+    // Iterate over the list of running standby containers, to find a standby resource that we have not already
+    // used for a failover for this active resoruce
+    for (String standbyContainerID : this.standbyContainerConstraints.get(activeContainerID)) {
+
+      if (samzaApplicationState.runningContainers.containsKey(standbyContainerID)) {
+        SamzaResource standbyContainerResource = samzaApplicationState.runningContainers.get(standbyContainerID);
```

Comment:
Can the state class expose this API?

SamzaResource getActiveContainerFor(SamzaResource r1). a nice thing is that you don't have to keep extracting specific fields and doing lookups